### PR TITLE
Use ES to map members of a dynamic group

### PR DESCRIPTION
**Pull request by dodobas**

### CODE REVIEW

File: temba/contacts/models.py [hunk 1]
```diff
@@ -2596,15 +2596,15 @@ def _get_dynamic_members(self):
...
+        from .search import contact_es_search, SearchException
+        from temba.utils.es import ES
         try:
-            qs, parsed = Contact.search(self.org, self.query)
+            es_search = contact_es_search(self.org, self.query, None).source(include=['id']).using(ES).scan()
 
-            return qs, parsed
+            return mapEStoDB(Contact, es_search), None
         except SearchException:  # pragma: no cover
             return Contact.objects.none(), None
```
Comment (nicpottier) [inline temba/contacts/models.py]: So I think there's a missing piece here from a lag perspective. We know ES can be at least 5 seconds behind, so we know some contacts will be missed. Seems to be absolutely correct we have to do something like:

1. Get the most recently modified contact in ES for our org
2. Do the ES search, get all members that match, that's our starting group
3. From the DB get all contacts that have a modified newer than the value in 1), evaluate their membership in the dynamic group ourselves, adding them or removing them from the ES set we got in step 2 appropriately
4. Return the final set

Its important that dynamic groups start off correct and that seems like the only way to leverage ES to accomplish that.

File: temba/contacts/models.py [hunk 2]
```diff
@@ -2577,20 +2577,40 @@ def reevaluate(self):
...
+        # remove contacts from the group that are not in the search
+        for members_chunk in chunk_list(to_remove_ids, 1000):
+            to_remove = Contact.objects.filter(id__in=members_chunk)
+
+            self.contacts.remove(*to_remove)
+
+            for changed_contact in to_remove:
+                changed_contact.handle_update(group=self)
+
+                # update group updated_at
```
Comment (nicpottier) [inline temba/contacts/models.py]: outside for

File: temba/contacts/models.py [hunk 3]
```diff
@@ -2599,14 +2619,51 @@ def _get_dynamic_members(self):
...
+            )
+        )
+
+        # check if contacts are members of the new group
+        for contact in db_contacts:
+            should_add = evaluate_query(self.org, self.query, contact_json=contact.as_search_json())
+
+            if should_add is True:
+                contact_ids.add(contact.id)
+
```
Comment (nicpottier) [inline temba/contacts/models.py]: sentry log if there are more than 1000 here?

File: temba/utils/models.py [hunk 4]
```diff
@@ -21,15 +21,18 @@ def generate_uuid():
...
     We use object PKs from ElasticSearch result set and select those objects in the database
     """
     pks = [result.id for result in es_queryset]
 
-    # TODO: order_by must be the same as the sort_by on ES, since we are losing the order of results
-    return model.objects.filter(id__in=pks).order_by('-modified_on').prefetch_related('org', 'all_groups')
+    if only_ids:
+        return pks
+    else:
+        # TODO: order_by must be the same as the sort_by on ES, since we are losing the order of results
```
Comment (rowanseymour) [inline temba/utils/models.py]: this isn't a TODO

**Pull request merged**